### ISSUE TITLE
[SR-15176] Fixit for `override func` that should be `override var`

### ISSUE THREAD

**Issue by CodaFi** [apple/swift#57499]
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-15176      |
   |Radar             | rdar://28337871         |
   |Original Reporter | @CodaFi      |
   |Type              | Improvement    |

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Compiler    |
   |Labels            | Improvement, DiagnosticsQoI, StarterBug        |
   |Assignee          | jiaren wang (JIRA)      |
   |Priority          | Medium      |

   

   md5: b17661d61829d04b65f822d78b07388a

  </details>

**Issue Description:**

Consider

```swift
class C {
  var canBecomeFirstResponder: Bool

  init() {}
}

class D: C {
    override func canBecomeFirstResponder() -> Bool {
    }
}

```

We should add a special case to `diagnoseGeneralOverrideFailure` that tries to see if it can correct no-args FuncDecls to VarDecls and vice-versa.

**Comment by RAJAGOPALAN-GANGADHARAN** [comment]
Hello @AnthonyLatsis -  is this issue still open, can I give it a shot? Thanks!

**Comment by AnthonyLatsis** [comment]
You're welcome to! Make sure it still reproduces though. Want we to assign you?

**Comment by danielfalbo** [comment]
I got busier and did not complete it yet, but I did make progress on this and was planning on opening the PR soon.

**Comment by Rajveer100** [comment]
Let's try this!

**Comment by Rajveer100** [comment]
Hey @LucianoPAlmeida @AnthonyLatsis , I managed to find the error call site:
File --> `/lib/Sema/TypeCheckDeclPrimary`

(Case 1: When the method is overridden instead of the property)
 
```
FD->diagnose(diag::method_does_not_override, isClassContext)
                .highlight(OA->getLocation());
```

(Case 2: When the property is overridden instead of the method)

```
VD->diagnose(diag::property_does_not_override, isClassContext)
                .highlight(OA->getLocation());
```

So I think fixing the actual issue remains. I will try to do it!

**Comment by Rajveer100** [comment]
@AnthonyLatsis I have a little off-topic issue, I have come across this earlier also, what I did was to delete my workspace and again create it, but is there any other solution to this?

`<unknown>:0: error: compiled module was created by an older version of the compiler; rebuild 'Swift' and try again:`

**Comment by Rajveer100** [comment]
PS: Rebuilding fixed the issue!

**Comment by AnthonyLatsis** [comment]
> is there any other solution to this?

This means the [binary module format](https://github.com/apple/swift/blob/main/docs/ABIStabilityManifesto.md#the-big-picture) version associated with the just-built compiler is different than the one that was used to serialize the Standard Library. Most likely, you pulled in the latest changes and rebuilt just the compiler (`swift-frontend`) rather than everything, including the Standard Library. Re-creating the workspace is almost never the answer, because the workspace or the generated Xcode projects are not responsible for building anything — the Ninja build is.

**Comment by Rajveer100** [comment]
Yes, also one more thing, I am suddenly encountering problems with `colouring` and `type hints` not working properly in Xcode files while I am making changes, I tried clearing the developers derived data folder too, which could potentially re index the files, but this isn't still working...

**Comment by AnthonyLatsis** [comment]
I have always seen these functional issues come up randomly when navigating the code base in Xcode. They usually get worse the older and more outdated the generated project becomes. Re-generating the projects after every checkout is probably the right answer, but I wouldn’t know for certain: I am too lazy to always keep the generated projects in sync, and tend to fall back to searching or spawning tabs whenever these issues show up.

**Comment by Rajveer100** [comment]
Ok I managed to fix the problem with some workaround, finally!

**Comment by Rajveer100** [comment]
Actually in my case, it was a different issue, so actually if you need a detailed explanation, here you go:

Basically, due to some of my older instances of projects, when I built the Xcode project, randomly the old instance of workspace and project would get created, so as soon as it got created, I deleted it, I also cleared the derived data, and finally after re launching Xcode, the hang error happens, this really made me happy, as I was aware of this one, which can be easily fixed (this was the first part of fix), by deleting autosave information, next I re launched and then everything was working again...

So actually the initial error which I said (related to syntax colouring, code completion, features like jump to definition), which weren't working, I noticed from that part, when only some of the functions weren't able to jump to original file, then, I realised I must check in the Xcode's console log, where I found the bug, Xcode wasn't able to find the files itself related to some dependencies...

So overall, if anyone encounters such issues in the future, follow the below steps:

1. Quit Xcode
2. Delete Derived data from Developers folder
3. Re launch Xcode
4. If the issue persists, build the project again
5. Then check if in the relative path (where your workspace was created), any extra files randomly pop there, delete them.
6. Quit Xcode

Now, you will see this nice hanging error (which is actually a good sign, that the problem is fixed)
All you need to do now is delete AutoSave information in the Library folder, quit and re launch, you should be fine.

**Comment by Rajveer100** [comment]
@AnthonyLatsis I awkwardly thought it as an issue, it wasn't an issue, actually when I was looking for type hint to get this `FD->getResultInterfaceType()->getRValueType()`, I typed `FD->getInterfaceType()`, and was expecting to show a type hint for it. Such a weird mistake I made, this made me think something was wrong with my work area, anyway.

**Comment by Rajveer100** [comment]
Also, are there any `docs` related to these specific types of files, meaning, learning about the `Types`, `Decl`, etc, to grab the pointers for exactly what I am looking for, or do I have to manually learn and figure out?

**Comment by Rajveer100** [comment]
Like in this issue, I will have to find a match for the function parameter and its declaration in the parent class, to correctly identify whether this override can be fixed...same goes for the property...

**Comment by LucianoPAlmeida** [comment]
> Like in this issue, I will have to find a match for the function parameter and its declaration in the parent class, to correctly identify whether this override can be fixed...same goes for the property...

There is a little bit of info [here](https://forums.swift.org/t/hint-at-how-to-get-properties-of-inherited-type/59632)

**Comment by Rajveer100** [comment]
Hey thanks a lot @LucianoPAlmeida! Looks like Swift Forums is the go for asking such questions!

**Comment by LucianoPAlmeida** [comment]
There is a [compiler](https://forums.swift.org/c/development/compiler/9) category on the forums where you can ask more general questions about development in the compiler. 
But if those are related to the issue is fine to keep the discussion here or on PRs =]

**Comment by AnthonyLatsis** [comment]
> actually when I was looking for type hint 

What do you mean by "type hint"? Code completion hints?

> Also, are there any `docs` related to these specific types of files, meaning, learning about the `Types`, `Decl`, etc, to grab the pointers for exactly what I am looking for, or do I have to manually learn and figure out?

I do not know of any detailed write-ups on the AST design or the parts of semantic analysis that do not involve subtle and complex algorithms among our docs. This is where the inline documentation and comments come in handy. But we do have a list of [external resources](https://github.com/apple/swift/blob/main/docs/ExternalResources.md), many of which I found very helpful back in the day, with many more yet to read. Check out the article by Slava Pestov about how stuff is represented in the compiler.

**Comment by Rajveer100** [comment]
Thanks! @AnthonyLatsis 
I will go through the external resources!

And yeah I meant code competition and syntax colouring, I had fixed it by changing the target, it got set to a different one, maybe I had check marked by mistake while editing code or setting breakpoints due to which only files related to those dependencies would show up in the colouring and code competition, other lines(in the same file or different one) would be white, and the jump to definition and other related features wouldn't work, which I misinterpreted that it was due to some other issue in my code base. 

Why was this issue closed and re opened?

**Comment by Rajveer100** [comment]
I have added the notes in the diagnostic file, I hope the changes are relatively fine:
[Commits](https://github.com/Rajveer100/swift/tree/branch-for-issue-57499)

**Comment by AnthonyLatsis** [comment]
Have you tested them? Please open a pull request if you want your changes to be reviewed.

**Comment by Rajveer100** [comment]
Coming soon!

**Comment by Rajveer100** [comment]
I have one question, regarding the `super class inherit`, should we consider the case for only one base class or can it be multiple classes, like, for example, `class D` could inherit from two classes `B` and `C`, where the function could be from either of them...

**Comment by Rajveer100** [comment]
I finally managed to do it! I just need to provide a fixit now!

Here is the output:

```
/Users/rajveersingh/Desktop/swift-project/build/Project-Swift-Compiler/Debug-Area.swift:15:19: error: method does not override any method from its superclass
    override func canBecomeFirstResponder() -> Bool {
    ~~~~~~~~      ^
/Users/rajveersingh/Desktop/swift-project/build/Project-Swift-Compiler/Debug-Area.swift:15:19: note: did you mean to override the property 'canBecomeFirstResponder'?
    override func canBecomeFirstResponder() -> Bool {
                  ^
```

**Comment by AnthonyLatsis** [comment]
> I have one question, regarding the `super class inherit`, should we consider the case for only one base class or can it be multiple classes

Swift does not have multiple inheritance.

**Comment by Rajveer100** [comment]
Oops! My bad! I misinterpreted it with `C++` as it has the `'virtual' keyword` to handle the diamond problem.

**Comment by wesleykebrown** [comment]
Hi @CodaFi. I'm working in a full-stack development cohort with a requirement for an open source contribution. I would love to try this one as my first one as I have an interest in learning Swift.

**Comment by AnthonyLatsis** [comment]
@wesleykebrown Assigned you.
